### ISSUE TITLE
No object detection for panoramic photos Tensorflow

### ISSUE THREAD

**Issue by hashtriton** [tensorflow/tensorflow#41741]
Good day. I ask for your help.

I tried almost all the models, but I didn't get the desired result.

TF finds objects but only at close proximity, if the photo is panoramic, then the detection does not work or works extremely poorly 
![Аннотация 2020-07-26 090233](https://user-images.githubusercontent.com/32834586/88474754-59323b00-cf32-11ea-87bc-007034ddb758.png)
![Аннотация 2020-07-26 090334](https://user-images.githubusercontent.com/32834586/88474756-5afbfe80-cf32-11ea-905c-89420a937b48.png)
The marked images contain both panoramic and close-up photographs. About the same amount

```model {
  faster_rcnn {
    num_classes: 12
    image_resizer {
      keep_aspect_ratio_resizer {
        min_dimension: 600
        max_dimension: 1024
      }
    }
    feature_extractor {
      type: "faster_rcnn_resnet50"
      first_stage_features_stride: 16
    }
    first_stage_anchor_generator {
      grid_anchor_generator {
        height_stride: 16
        width_stride: 16
        scales: 0.25
        scales: 0.5
        scales: 1.0
        scales: 2.0
        aspect_ratios: 0.5
        aspect_ratios: 1.0
        aspect_ratios: 2.0
      }
    }
    first_stage_box_predictor_conv_hyperparams {
      op: CONV
      regularizer {
        l2_regularizer {
          weight: 0.0
        }
      }
      initializer {
        truncated_normal_initializer {
          stddev: 0.00999999977648
        }
      }
    }
    first_stage_nms_score_threshold: 0.0
    first_stage_nms_iou_threshold: 0.699999988079
    first_stage_max_proposals: 50
    first_stage_localization_loss_weight: 2.0
    first_stage_objectness_loss_weight: 1.0
    initial_crop_size: 14
    maxpool_kernel_size: 2
    maxpool_stride: 2
    second_stage_box_predictor {
      mask_rcnn_box_predictor {
        fc_hyperparams {
          op: FC
          regularizer {
            l2_regularizer {
              weight: 0.0
            }
          }
          initializer {
            variance_scaling_initializer {
              factor: 1.0
              uniform: true
              mode: FAN_AVG
            }
          }
        }
        use_dropout: false
        dropout_keep_probability: 1.0
      }
    }
    second_stage_post_processing {
      batch_non_max_suppression {
        score_threshold: 0.300000011921
        iou_threshold: 0.600000023842
        max_detections_per_class: 20
        max_total_detections: 20
      }
      score_converter: SOFTMAX
    }
    second_stage_localization_loss_weight: 2.0
    second_stage_classification_loss_weight: 1.0
    second_stage_batch_size: 49
  }
}
train_config {
  batch_size: 1
  data_augmentation_options {
    random_horizontal_flip {
    }
  }
  optimizer {
    momentum_optimizer {
      learning_rate {
        manual_step_learning_rate {
          initial_learning_rate: 0.000300000014249
          
          schedule {
            step: 900000
            learning_rate: 2.99999992421e-05
          }
          schedule {
            step: 1200000
            learning_rate: 3.00000010611e-06
          }
        }
      }
      momentum_optimizer_value: 0.899999976158
    }
    use_moving_average: false
  }
  gradient_clipping_by_norm: 10.0
  fine_tune_checkpoint: "C:/tensorflow1/models/research/object_detection/faster_rcnn_resnet50_lowproposals_coco_2018_01_28/model.ckpt"
  from_detection_checkpoint: true
  num_steps: 200000
}
train_input_reader {
  label_map_path: "C:/tensorflow1/models/research/object_detection/training/labelmap.pbtxt"
  tf_record_input_reader {
    input_path: "C:/tensorflow1/models/research/object_detection/train.record"
  }
}
eval_config {
  num_examples: 316
  max_evals: 10
  use_moving_averages: false
}
eval_input_reader {
  label_map_path: "C:/tensorflow1/models/research/object_detection/training/labelmap.pbtxt"
  shuffle: false
  num_readers: 1
  tf_record_input_reader {
    input_path: "C:/tensorflow1/models/research/object_detection/test.record"
  }
}
```

Object_detection_image.py
```
######## Image Object Detection Using Tensorflow-trained Classifier #########
# Import packages
import os
import cv2
import numpy as np
import tensorflow as tf
import sys

# This is needed since the notebook is stored in the object_detection folder.
sys.path.append("..")

# Import utilites
from utils import label_map_util
from utils import visualization_utils as vis_util

# Name of the directory containing the object detection module we're using
MODEL_NAME = 'inference_graph'
IMAGE_NAME = 'test1.jpeg'

# Grab path to current working directory
CWD_PATH = os.getcwd()

# Path to frozen detection graph .pb file, which contains the model that is used
# for object detection.
PATH_TO_CKPT = os.path.join(CWD_PATH,MODEL_NAME,'frozen_inference_graph.pb')

# Path to label map file
PATH_TO_LABELS = os.path.join(CWD_PATH,'training','labelmap.pbtxt')

# Path to image
PATH_TO_IMAGE = os.path.join(CWD_PATH,IMAGE_NAME)

# Number of classes the object detector can identify
NUM_CLASSES = 12

# Load the label map.
# Label maps map indices to category names, so that when our convolution
# network predicts `5`, we know that this corresponds to `king`.
# Here we use internal utility functions, but anything that returns a
# dictionary mapping integers to appropriate string labels would be fine
label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
category_index = label_map_util.create_category_index(categories)

# Load the Tensorflow model into memory.
detection_graph = tf.Graph()
with detection_graph.as_default():
    od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(PATH_TO_CKPT, 'rb') as fid:
        serialized_graph = fid.read()
        od_graph_def.ParseFromString(serialized_graph)
        tf.import_graph_def(od_graph_def, name='')

    sess = tf.Session(graph=detection_graph)

# Define input and output tensors (i.e. data) for the object detection classifier

# Input tensor is the image
image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')

# Output tensors are the detection boxes, scores, and classes
# Each box represents a part of the image where a particular object was detected
detection_boxes = detection_graph.get_tensor_by_name('detection_boxes:0')

# Each score represents level of confidence for each of the objects.
# The score is shown on the result image, together with the class label.
detection_scores = detection_graph.get_tensor_by_name('detection_scores:0')
detection_classes = detection_graph.get_tensor_by_name('detection_classes:0')

# Number of objects detected
num_detections = detection_graph.get_tensor_by_name('num_detections:0')

# Load image using OpenCV and
# expand image dimensions to have shape: [1, None, None, 3]
# i.e. a single-column array, where each item in the column has the pixel RGB value
image = cv2.imread(PATH_TO_IMAGE)
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image_expanded = np.expand_dims(image_rgb, axis=0)

# Perform the actual detection by running the model with the image as input
(boxes, scores, classes, num) = sess.run(
    [detection_boxes, detection_scores, detection_classes, num_detections],
    feed_dict={image_tensor: image_expanded})

# Draw the results of the detection (aka 'visulaize the results')

vis_util.visualize_boxes_and_labels_on_image_array(
    image,
    np.squeeze(boxes),
    np.squeeze(classes).astype(np.int32),
    np.squeeze(scores),
    category_index,
    use_normalized_coordinates=True,
    line_thickness=2,
    min_score_thresh=0.30)

# All the results have been drawn on image. Now display the image.
cv2.imshow('Object detector', image)

# Press any key to close the image
cv2.waitKey(0)

# Clean up
cv2.destroyAllWindows()
```
![tvorog_dop_pon (80)](https://user-images.githubusercontent.com/32834586/88474783-95659b80-cf32-11ea-8bb2-e3ea0047fa95.jpeg)
![tvorog_dop_pon (82)](https://user-images.githubusercontent.com/32834586/88474786-9991b900-cf32-11ea-846b-534eef0262f0.jpeg)
![tvorog_dop_sr (3)](https://user-images.githubusercontent.com/32834586/88474797-add5b600-cf32-11ea-8671-3b31e656e2d3.jpeg)

I have now copied the panoramic photo to the training folder in the folder for object_detection. No matches found that is, a completely identical image from the training set does not find a match

**Comment by jpodivin** [comment]
There are two simple possibilities: 

1. Preprocessing. How does the panoramic image actually 'look' when passed to mode? It might be distorted. 
2. It is possible that the model was trained only on small images. In that case it might not recognize objects on larger images, because it contains different features. Or rather does not contain expected features.

**Comment by hashtriton** [comment]
Good day. I did not understand about point 1.
Point 2: No, the training kit also includes panoramic and close-up photos. That is why I have attached sample photos

**Comment by hashtriton** [comment]
> There are two simple possibilities:
> 
> 1. Preprocessing. How does the panoramic image actually 'look' when passed to mode? It might be distorted.
> 2. It is possible that the model was trained only on small images. In that case it might not recognize objects on larger images, because it contains different features. Or rather does not contain expected features.

Good day. I did not understand about point 1.
Point 2: No, the training kit also includes panoramic and close-up photos. That is why I have attached sample photos

**Comment by jpodivin** [comment]
About point 1: 
When you pass an input to the model it can be transformed in various ways, depending on how you set up your pipeline. 
Resolution, colors, cropping etc. it is possible that your evaluation data undergoes different transformation than your training data, if that's the case your model will have trouble recognizing the features it learned on training data. 

Anyway, I do recommend that you look at the raw output of your model: `boxes, scores, classes, num`. It will tell you more than the resulting bounding boxes.

**Comment by hashtriton** [comment]
> О пункте 1:
> Когда вы передаете входные данные в модель, они могут быть преобразованы различными способами, в зависимости от того, как вы настроили свой конвейер.
> Разрешение, цвета, обрезка и т. Д. Возможно, что ваши оценочные данные претерпевают иное преобразование, чем ваши тренировочные данные, если в этом случае вашей модели будет трудно распознать особенности, которые она изучила на тренировочных данных.
> 
> Во всяком случае, я рекомендую вам посмотреть на сырье выхода модели: `boxes, scores, classes, num`. Он расскажет вам больше, чем получившиеся ограничивающие рамки.

Thanks for your reply. Yes, I understand what you mean.
However, I am using the same .config to train and find objects. I have attached the code for this config

**Comment by hashtriton** [comment]
> About point 1:
> When you pass an input to the model it can be transformed in various ways, depending on how you set up your pipeline.
> Resolution, colors, cropping etc. it is possible that your evaluation data undergoes different transformation than your training data, if that's the case your model will have trouble recognizing the features it learned on training data.
> 
> Anyway, I do recommend that you look at the raw output of your model: `boxes, scores, classes, num`. It will tell you more than the resulting bounding boxes.

Thanks for your reply. Yes, I understand what you mean.
However, I am using the same .config to train and find objects. I have attached the code for this config

**Comment by hashtriton** [comment]
> About point 1:
> When you pass an input to the model it can be transformed in various ways, depending on how you set up your pipeline.
> Resolution, colors, cropping etc. it is possible that your evaluation data undergoes different transformation than your training data, if that's the case your model will have trouble recognizing the features it learned on training data.
> 
> Anyway, I do recommend that you look at the raw output of your model: `boxes, scores, classes, num`. It will tell you more than the resulting bounding boxes.

Ok, 
Could you give your script for finding objects?

**Comment by hashtriton** [comment]
My object_detection.py 
```
######## Image Object Detection Using Tensorflow-trained Classifier #########
#
# Author: Evan Juras
# Date: 1/15/18
# Description: 
# This program uses a TensorFlow-trained neural network to perform object detection.
# It loads the classifier and uses it to perform object detection on an image.
# It draws boxes, scores, and labels around the objects of interest in the image.

## Some of the code is copied from Google's example at
## https://github.com/tensorflow/models/blob/master/research/object_detection/object_detection_tutorial.ipynb

## and some is copied from Dat Tran's example at
## https://github.com/datitran/object_detector_app/blob/master/object_detection_app.py

## but I changed it to make it more understandable to me.

# Import packages
import os
import cv2
import numpy as np
import tensorflow as tf
import sys

# This is needed since the notebook is stored in the object_detection folder.
sys.path.append("..")

# Import utilites
from utils import label_map_util
from utils import visualization_utils as vis_util

# Name of the directory containing the object detection module we're using
MODEL_NAME = 'inference_graph'
IMAGE_NAME = 'test1.jpeg'

# Grab path to current working directory
CWD_PATH = os.getcwd()

# Path to frozen detection graph .pb file, which contains the model that is used
# for object detection.
PATH_TO_CKPT = os.path.join(CWD_PATH,MODEL_NAME,'frozen_inference_graph.pb')

# Path to label map file
PATH_TO_LABELS = os.path.join(CWD_PATH,'training','labelmap.pbtxt')

# Path to image
PATH_TO_IMAGE = os.path.join(CWD_PATH,IMAGE_NAME)

# Number of classes the object detector can identify
NUM_CLASSES = 12

# Load the label map.
# Label maps map indices to category names, so that when our convolution
# network predicts `5`, we know that this corresponds to `king`.
# Here we use internal utility functions, but anything that returns a
# dictionary mapping integers to appropriate string labels would be fine
label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
category_index = label_map_util.create_category_index(categories)

# Load the Tensorflow model into memory.
detection_graph = tf.Graph()
with detection_graph.as_default():
    od_graph_def = tf.GraphDef()
    with tf.gfile.GFile(PATH_TO_CKPT, 'rb') as fid:
        serialized_graph = fid.read()
        od_graph_def.ParseFromString(serialized_graph)
        tf.import_graph_def(od_graph_def, name='')

    sess = tf.Session(graph=detection_graph)

# Define input and output tensors (i.e. data) for the object detection classifier

# Input tensor is the image
image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')

# Output tensors are the detection boxes, scores, and classes
# Each box represents a part of the image where a particular object was detected
detection_boxes = detection_graph.get_tensor_by_name('detection_boxes:0')

# Each score represents level of confidence for each of the objects.
# The score is shown on the result image, together with the class label.
detection_scores = detection_graph.get_tensor_by_name('detection_scores:0')
detection_classes = detection_graph.get_tensor_by_name('detection_classes:0')

# Number of objects detected
num_detections = detection_graph.get_tensor_by_name('num_detections:0')

# Load image using OpenCV and
# expand image dimensions to have shape: [1, None, None, 3]
# i.e. a single-column array, where each item in the column has the pixel RGB value
image = cv2.imread(PATH_TO_IMAGE)
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
image_expanded = np.expand_dims(image_rgb, axis=0)

# Perform the actual detection by running the model with the image as input
(boxes, scores, classes, num) = sess.run(
    [detection_boxes, detection_scores, detection_classes, num_detections],
    feed_dict={image_tensor: image_expanded})

# Draw the results of the detection (aka 'visulaize the results')

vis_util.visualize_boxes_and_labels_on_image_array(
    image,
    np.squeeze(boxes),
    np.squeeze(classes).astype(np.int32),
    np.squeeze(scores),
    category_index,
    use_normalized_coordinates=True,
    line_thickness=2,
    min_score_thresh=0.30)

# All the results have been drawn on image. Now display the image.
cv2.imshow('Object detector', image)

# Press any key to close the image
cv2.waitKey(0)

# Clean up
cv2.destroyAllWindows()
```

**Comment by Kira2288** [comment]
Try Feature Pyramid Network on top of Faster R-CNN. They are good at recognizing objects at vastly different scales.

**Comment by hashtriton** [comment]
> Try Feature Pyramid Network on top of Faster R-CNN. They are good at recognizing objects at vastly different scales.

Thanks for your reply. Could you give a link where I can study about this? Desirable with a real example?

I understand correctly that you think that the problem is with the model, not with something else

**Comment by Kira2288** [comment]
> > Try Feature Pyramid Network on top of Faster R-CNN. They are good at recognizing objects at vastly different scales.
> 
> Thanks for your reply. Could you give a link where I can study about this? Desirable with a real example?
> 
> I understand correctly that you think that the problem is with the model, not with something else

https://medium.com/@jonathan_hui/understanding-feature-pyramid-networks-for-object-detection-fpn-45b227b9106c

https://github.com/DetectionTeamUCAS/FPN_Tensorflow

**Comment by hashtriton** [comment]
> > > Try Feature Pyramid Network on top of Faster R-CNN. They are good at recognizing objects at vastly different scales.
> > 
> > 
> > Thanks for your reply. Could you give a link where I can study about this? Desirable with a real example?
> > I understand correctly that you think that the problem is with the model, not with something else
> 
> https://medium.com/@jonathan_hui/understanding-feature-pyramid-networks-for-object-detection-fpn-45b227b9106c
> 
> https://github.com/DetectionTeamUCAS/FPN_Tensorflow

Thank you so much. I started training, I'm testing.

Question: I have a photo where I mark one item, but there is another item on it that I have not marked. Could this affect the bottom line?

In other words, I can have 2 identical photos where the first one is marked with PRODUCT1, and the second with PRODUCT2

**Comment by Kira2288** [comment]
You should mark all the product you are detecting because if your network finds some similar product in the picture which are not marked but should be then the weights will adjust accordingly and accuracy may decrease.

**Comment by gowthamkpr** [comment]
@hashtriton Has this issue been resolved?
